src/mpi/misc/memory_alloc_kinds.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+static const char *memory_alloc_kinds[3][5] = {
+    /* mpi 4.1 kinds */
+    {"mpi", "alloc_mem", "win_allocate", "win_allocate_shared", NULL},
+    {"system", NULL},
+    {NULL}
+};
+
+static bool is_supported(const char *kind);
+
+int MPIR_get_supported_memory_kinds(char *requested_kinds, char **out_kinds)
+{
+    char *tmp_strs[32];
+    int num = 2;
+    /* mpi and system kinds are always supported */
+    tmp_strs[0] = MPL_strdup("mpi");
+    tmp_strs[1] = MPL_strdup("system");
+
+    /* match user kinds with supported kinds */
+    if (requested_kinds != NULL) {
+        char *tmp = MPL_strdup(requested_kinds);
+        char *save_tmp = tmp;
+        for (char *kind = MPL_strsep(&tmp, ","); kind; kind = MPL_strsep(&tmp, ",")) {
+            if (!MPL_stricmp(kind, "mpi") || !MPL_stricmp(kind, "system")) {
+                continue;
+            } else if (is_supported(kind)) {
+                tmp_strs[num++] = MPL_strdup(kind);
+                /* FIXME: use dynamic storage */
```

Comment:
I think 32 is too small. How about 1024?

Reply:
Sure. Will push an update.